certstore_linux.go

```diff
@@ -1,14 +1,318 @@
+		return nil, errors.New("failure to allocate memory for digest")
+	}
+	defer C.SECITEM_FreeItem(digestC, 1)
+
+	resultC := C.SECITEM_AllocItem(nil, nil, C.uint(C.PK11_SignatureLen(key)))
+	if resultC == nil {
+		return nil, errors.New("failure to allocate memory for signature")
+	}
+	defer C.SECITEM_FreeItem(resultC, 1)
+
+	digestC.len = C.uint(len(digest))
+	C.memcpy(unsafe.Pointer(digestC.data), unsafe.Pointer(&digest[0]), C.size_t(len(digest)))
+
+	if C.SGN_Digest(key, goToNssAlg[hash], resultC, digestC) != C.SECSuccess {
+		if C.PORT_GetError() == C.SEC_ERROR_INVALID_ALGORITHM {
+			return nil, ErrUnsupportedHash
+		}
+		return nil, fmt.Errorf("error signing: %s\n", C.GoString(C.GetErrorString()))
+	}
+
+	if key.keyType == C.ecKey {
+		var dsaSig C.SECItem
+		if C.DSAU_EncodeDerSigWithLen(&dsaSig, resultC, resultC.len) != C.SECSuccess {
+			return nil, errors.New("unable to generate dsa signature")
+		}
+		resultC = &dsaSig
+	}
+	return C.GoBytes(unsafe.Pointer(resultC.data), C.int(resultC.len)), nil
+}
+
+// Import implements the Store interface.
+func (nssStore) Import(data []byte, password string) error {
+	unicodePassword, err := bmpString(password)
+	if unicodePassword == nil {
+		return err
+	}
+	pass := C.SECITEM_AllocItem(nil, nil, C.uint(len(unicodePassword)))
+	if pass == nil {
+		return errors.New("SECITEM_AllocItem failed")
+	}
+	defer C.SECITEM_FreeItem(pass, 1)
+	C.memcpy(unsafe.Pointer(pass.data), unsafe.Pointer(&unicodePassword[0]), C.size_t(len(unicodePassword)))
+	var (
+		p12     = C.SEC_PKCS12DecoderStart(pass, nil, nil, nil, nil, nil, nil, nil)
+		decoded = C.SEC_PKCS12DecoderUpdate(p12, (*C.uchar)(unsafe.Pointer(&data[0])), C.size_t(len(data)))
+	)
+	if decoded != 0 {
+		return fmt.Errorf("error %d, P12 decoding failed", int(C.PR_GetError()))
+	}
+	authenticated := C.SEC_PKCS12DecoderVerify(p12)
+	if authenticated != 0 {
+		return fmt.Errorf("error %d, P12 authentication failed", int(C.PR_GetError()))
+	}
+	validated := C.SEC_PKCS12DecoderValidateBags(p12, (*[0]byte)(C.P12U_NicknameCollisionCallback))
+	if validated != 0 {
+		return fmt.Errorf("error %d, P12 validation failed", int(C.PR_GetError()))
+	}
+	imported := C.SEC_PKCS12DecoderImportBags(p12)
+	if imported != 0 {
+		return fmt.Errorf("error %d, P12 import failed", int(C.PR_GetError()))
+	}
+	return nil
+}
+
+// Close implements the Store interface.
+func (ctx nssStore) Close() {
+	nssCtx := C.NSSInitContext(ctx)
+	C.NSS_ShutdownContext(&nssCtx)
+}
+
+// openStore opens the current user's NSS database.
 func openStore() (Store, error) {
-	return nil, errors.New("certstore only works on macOS and Windows")
+	homeDir, err := os.UserHomeDir()
+
+	if err != nil {
+		return nil, err
+	}
+
+	nssdb := path.Join(homeDir, ".pki", "nssdb")
+	if _, err := os.Stat(nssdb); os.IsNotExist(err) {
+		return nil, fmt.Errorf("NSS database not found at %s", nssdb)
+	}
+
+	ctx := C.NSS_InitContext(C.CString(nssdb), C.CString(""), C.CString(""), C.CString(""), nil, C.uint(0))
```

Comment:
Thanks for switching to `NSS_InitContext `. This is feels _much_ nicer than using the global init/shutdown.